InputText, InputTextMultiline, InputFloat, InputInt

supudo opened the pull request:
Hey :),

This one contains the InputXYZ functions. There's one not-so-minor issue, which also is present in the ImGui demo window.
For InputText if the buffer size is not specified as something big like 1024 or bigger it will take the passed value and use it as a limit - so if you pass "Text" it will display text but will also put limit of 4 chars on the input. No idea why this happens, so i've put a 2056 char array as limit for now. Same applies for the InputTextMultiline function.
Another issue with the inputs i noticed is that it works too fast, meaning if i type "A" it's ok, but if i do backspace (delete) and enter (new lines) it deletes multiple characters/words or puts lots of new lines. Same behaviour exists in the ImGui demo window by the way, so i think it might be something either in the implementation or in Cython itself.

Will keep digging to see if i can fix it.

P.S. Same type of issues i noticed with scrolling - if you have a big height window and you scroll with the mouse for ex., it always goes either at the top or at the bottom, you can't position it in the middle. Same issues with the combo lists. 

P.P.S. All of those are observed on Python 3.5 with the PySDL implementation, so they might not be present in Python 2.X.

Cheers,
S.

swistakm commented:
I will pull it locally to see what may be the cause if these issues.

If something does not work with imgui demo window it is either a bug in imgui C++ implementation or a problem with the inputs processing (so PySDL/GLFW integration). I will review the #14 PySLD2 PR soon and then compare results between different integrations.

swistakm commented on imgui/core.pyx:
```diff
@@ -3273,6 +3293,475 @@ def drag_int4(
+    # return changed, bytes_text.decode('utf-8')
+
+    cdef bytes bytes_text = _bytes(value)
+    cdef char inout_text[2056]
```
This `inout_text` buffer size should be a function argument. This will require dynamic allocation and proper string type handling via Python/C API.

swistakm commented on imgui/core.pyx:
```diff
@@ -3273,6 +3293,475 @@ def drag_int4(
+
+    cdef bytes bytes_text = _bytes(value)
+    cdef char inout_text[2056]
+    strcpy(inout_text, bytes_text)
```
Could be simply `strcpy(inout_text, _bytes(value))`. The local `bytes_text` variable is redundant (see comments below).

swistakm commented on imgui/core.pyx:
```diff
@@ -3273,6 +3293,475 @@ def drag_int4(
+        _bytes(label), inout_text, sizeof(inout_text), flags, NULL, NULL
+    )
+
+    return changed, bytes_text.decode('utf-8')
```
We need to return a buffer, so `inout_text ` instead of `bytes_text`. Also I'm not sure about this `decode('utf-8')` this introduces a compatibility issue in Python2.7.

In Python2.7 this function will accept `str` as `value` but returns `unicode`. We need to either force `str` somehow. Maybe right now for consistency we need some opposite implementation of the `_bytes()` function. Maybe `_from_bytes()`?

swistakm commented on imgui/core.pyx:
```diff
@@ -3273,6 +3293,475 @@ def drag_int4(
+
+        text_val = 'Please, type the coefficient here.'
+        imgui.begin("Example: text input")
+        changed, text_val = imgui.input_text('Amount:', text_val)
```
Label will be displayed on right so semicolon looks weird in the example.
You can either use hidden label with `##` prefix:
```python
changed, text_val = imgui.input_text('##Amount:', text_val)
```
or simply use `imgui.same_line()`.

swistakm commented on imgui/core.pyx:
```diff
@@ -3273,6 +3293,475 @@ def drag_int4(
+        )
+    """
+
+    # return changed, bytes_text.decode('utf-8')
```
unnecessary comment

swistakm commented on doc/source/guide/inputtext-flags.rst:
```diff
@@ -0,0 +1,28 @@
+.. _guide-inputtext-flags:
+
+Using input text flags
+==================
```
Make the underline same width as the title. Otherwise Sphinx will generate warnings.

swistakm commented on imgui/core.pyx:
```diff
@@ -3273,6 +3293,475 @@ def drag_int4(
+    float width=0,
+    float height=0,
+    cimgui.ImGuiInputTextFlags flags=0
+):
```
Same comments as for `input_text()` function apply also here.

swistakm commented:
I did run thin in my local environment and here is my feedback:
1. The buffer size unfortunately must be set explicitly right now as a function parameter. Imgui expects that user knows how much of input he wants to process. Setting this to arbitrarily high value is not a proper solution. I would rather like to not have a default value for such param either because if we implement #8 fully, this will have to be a keyword-only argument.
2.  Remember that buffer size is in bytes and the input is in unicode. This means that buffer size does not equal the number of characters. Function docstring should have a clear note on that.
3. I have tried to run example code in both GLFW2 implementation and PySDL2. I have noticed the "multiple backspace and newline characters" issue only on GLFW3 and Python 3. I don't know what is the problem yet. But since it happens also in the demo window it is just implementation-specific issue or internal imgui problem. Will have to try raw C++ example from imgui sources.
4. I have found that `PySDL2` does not handle key modifiers correctly. Pressing SHIFT, CTRL, ALT etc. yields semi-random unicode code points.
5. Callbacks are missing but they are optional so I can add support for them in the future.

3., 4. and 5. can be handled in separate issues/PRs.

swistakm commented:
I have found the reason for problem 3. It is a timing issue and both PySDL2 and GLFW3 do not measure time deltas properly. Will be very easy to fix.

supudo commented:
Hey :),

All those PR got me confused and i won't be surprised if i messed the branches again.
Anyway, i've added the _from_bytes() (think it's correct) and fixed the input_text methods to be with length parameters.

Cheers,
S.

swistakm commented:
This branch is fine. You did not mess anything here :). The `_from_bytes()` function loos great. It is something that I exactly had on my mind.

There are just some two issues with `input_float3` and `input_float4` functions. You can find details in Appveyor build logs. Seems that creating this tests pipeline did finally pay out 🍾
